src/CloudNative.CloudEvents.Amqp/AmqpCloudEventMessage.cs

```diff
@@ -1,90 +1,96 @@
+                properties.Add(propKey, propValue);
             }
-            else if (cloudEvent.Data is string)
-            {
-                this.BodySection = new AmqpValue() { Value = cloudEvent.Data };
-            }
-
-            this.Properties = new Properties() { ContentType = cloudEvent.DataContentType?.MediaType };
-            this.ApplicationProperties = new ApplicationProperties();
-            MapHeaders(cloudEvent);
         }
 
-        void MapHeaders(CloudEvent cloudEvent)
+        /// <summary>
+        /// Convert data into a suitable format for inclusion in an Avro record.
+        /// TODO: Asynchronous version?
+        /// </summary>
+        /// <param name="data"></param>
+        /// <returns></returns>
+        private static RestrictedDescribed SerializeData(object data)
         {
-            foreach (var attribute in cloudEvent.GetAttributes())
+            switch (data)
             {
-                if (!attribute.Key.Equals(CloudEventAttributes.DataAttributeName(cloudEvent.SpecVersion)) &&
-                    !attribute.Key.Equals(CloudEventAttributes.DataContentTypeAttributeName(cloudEvent.SpecVersion)))
-                {
-                    string key = "cloudEvents:" + attribute.Key;
-                    if (attribute.Value is Uri)
-                    {
-                        this.ApplicationProperties.Map.Add(key, attribute.Value.ToString());
-                    }
-                    else if (attribute.Value is DateTimeOffset dto)
-                    {
-                        // AMQPNetLite doesn't support DateTimeOffset values, so convert to UTC.
-                        // That means we can't roundtrip events with non-UTC timestamps, but that's not awful.
-                        this.ApplicationProperties.Map.Add(key, dto.UtcDateTime);
-                    }
-                    else if (attribute.Value is string)
-                    {
-                        this.ApplicationProperties.Map.Add(key, attribute.Value);
-                    }
-                    else
-                    {
-                        Map dict = new Map();
-                        foreach (PropertyDescriptor descriptor in TypeDescriptor.GetProperties(attribute.Value))
-                        {
-                            dict.Add(descriptor.Name, descriptor.GetValue(attribute.Value));
-                        }
-
-                        this.ApplicationProperties.Map.Add("cloudEvents:" + attribute.Key, dict);
-                    }
-                }
+                case null:
+                    return null;
+                case byte[] bytes:
+                    return new Data { Binary = bytes };
+                case MemoryStream memoryStream:
+                    // Note: this will return the whole stream, regardless of position...
+                    return new Data { Binary = memoryStream.ToArray() };
+                case Stream stream:
+                    var buffer = new MemoryStream();
```

Comment:
Should the MemoryStream be disposed?

Reply:
No need - it wouldn't actually do anything.

Reply:
fair enough